File: examples/scripts/nvd_ingest.ts
#!/usr/bin/env bun

/**
 * NVD Support Car Ingestion Client (TypeScript/Bun)
 *
 * This TypeScript client provides functionality for ingesting metagenomic analysis
 * results into the NVD Support Car service. It is designed to run with Bun, which
 * provides fast startup times and built-in TypeScript support without requiring
 * compilation.
 *
 * The script handles GOTTCHA2 and STAST data formats, manages authentication via
 * bearer tokens, and automatically compresses data using gzip before transmission.
 * It can process large result files in batches to avoid memory issues and provides
 * progress feedback during ingestion.
 *
 * Installation of Bun:
 *   curl -fsSL https://bun.sh/install | bash
 *
 * Usage:
 *   bun run nvd_ingest.ts --help
 *   bun run nvd_ingest.ts gottcha2 --input results.tsv --sample-id SAMPLE001
 *   bun run nvd_ingest.ts stast --input blast.tsv --sample-id SAMPLE001 --task megablast
 *
 * Environment variables:
 *   NVD_SUPPORT_URL: Base URL of the NVD Support Car service
 *   NVD_BEARER_TOKEN: Bearer token for authentication
 *
 * When used in Nextflow pipelines, this script can be placed in the bin/ directory
 * and will be automatically available to all processes. The script's shebang line
 * ensures it will be executed with Bun when called directly.
 */

import { createReadStream, existsSync } from "node:fs";
import process from "node:process";
import readline from "node:readline";

// Type definitions for our data structures
interface Gottcha2Record {
	sample_id: string;
	level: string;
	name: string;
	taxid: string;
	read_count: number;
	total_bp_mapped: number;
	ani_ci95: number;
	covered_sig_len: number;
	best_sig_cov: number;
	depth: number;
	rel_abundance: number;
}

interface StastRecord {
	task: string;
	sample_id: string;
	qseqid: string;
	qlen: number;
	sseqid: string;
	stitle: string;
	length: number;
	pident: number;
	evalue: number;
	bitscore: number;
	sscinames: string;
	staxids: string;
	rank: string;
}

class IngestionClient {
	private baseUrl: string;
	private bearerToken: string;
	private headers: Record<string, string>;

	constructor(baseUrl: string, bearerToken: string) {
		this.baseUrl = baseUrl.replace(/\/$/, "");
		this.bearerToken = bearerToken;
		this.headers = {
			Authorization: `Bearer ${bearerToken}`,
			"Content-Type": "application/gzip",
			"Content-Encoding": "gzip",
		};
	}

	async sendRecords<T>(endpoint: string, records: T[]): Promise<boolean> {
		// Convert records to JSONL
		const jsonlData = records
			.map((record) => JSON.stringify(record))
			.join("\n");

		// Compress with gzip
		const encoder = new TextEncoder();
		const uncompressed = encoder.encode(jsonlData);
		const compressed = Bun.gzipSync(uncompressed);

		try {
			const response = await fetch(`${this.baseUrl}/${endpoint}`, {
				method: "POST",
				headers: this.headers,
				body: compressed,
			});

			if (!response.ok) {
				const errorText = await response.text();
				console.error(`HTTP error ${response.status}: ${errorText}`);
				return false;
			}

			console.log(
				`Successfully ingested ${records.length} records to ${endpoint}`,
			);
			return true;
		} catch (error) {
			console.error(`Error sending data: ${error}`);
			return false;
		}
	}

	async checkHealth(): Promise<boolean> {
		try {
			const response = await fetch(`${this.baseUrl}/healthz`, {
				headers: {
					Authorization: `Bearer ${this.bearerToken}`,
				},
			});

			if (response.ok) {
				const text = await response.text();
				console.log(`Service is healthy: ${text}`);
				return true;
			} else {
				console.error(`Health check failed with status ${response.status}`);
				return false;
			}
		} catch (error) {
			console.error(`Health check error: ${error}`);
			return false;
		}
	}
}

async function parseGottcha2TSV(
	filepath: string,
	sampleId: string,
): Promise<Gottcha2Record[]> {
	const records: Gottcha2Record[] = [];
	const fileStream = createReadStream(filepath);
	const rl = readline.createInterface({
		input: fileStream,
		crlfDelay: Infinity,
	});

	let lineNum = 0;
	for await (const line of rl) {
		lineNum++;

		// Skip header
		if (lineNum === 1 && line.startsWith("LEVEL")) {
			continue;
		}

		const parts = line.trim().split("\t");
		if (parts.length < 10) {
			console.warn(`Warning: Skipping line ${lineNum}, insufficient columns`);
			continue;
		}

		try {
			const record: Gottcha2Record = {
				sample_id: sampleId,
				level: parts[0],
				name: parts[1],
				taxid: parts[2],
				read_count: parseInt(parts[3], 10),
				total_bp_mapped: parseInt(parts[4], 10),
				ani_ci95: parseFloat(parts[5]),
				covered_sig_len: parseInt(parts[6], 10),
				best_sig_cov: parseFloat(parts[7]),
				depth: parseFloat(parts[8]),
				rel_abundance: parseFloat(parts[9]),
			};
			records.push(record);
		} catch (error) {
			console.warn(`Warning: Error parsing line ${lineNum}: ${error}`);
		}
	}

	return records;
}

async function parseStastTSV(
	filepath: string,
	sampleId: string,
	task: string,
): Promise<StastRecord[]> {
	const records: StastRecord[] = [];
	const fileStream = createReadStream(filepath);
	const rl = readline.createInterface({
		input: fileStream,
		crlfDelay: Infinity,
	});

	let lineNum = 0;
	for await (const line of rl) {
		lineNum++;

		// Skip header
		if (lineNum === 1 && line.startsWith("qseqid")) {
			continue;
		}

		const parts = line.trim().split("\t");
		if (parts.length < 11) {
			console.warn(`Warning: Skipping line ${lineNum}, insufficient columns`);
			continue;
		}

		try {
			const record: StastRecord = {
				task: task,
				sample_id: sampleId,
				qseqid: parts[0],
				qlen: parseInt(parts[1], 10),
				sseqid: parts[2],
				stitle: parts[3],
				length: parseInt(parts[4], 10),
				pident: parseFloat(parts[5]),
				evalue: parseFloat(parts[6]),
				bitscore: parseFloat(parts[7]),
				sscinames: parts[8],
				staxids: parts[9],
				rank: parts[10],
			};
			records.push(record);
		} catch (error) {
			console.warn(`Warning: Error parsing line ${lineNum}: ${error}`);
		}
	}

	return records;
}

async function ingestGottcha2(
	client: IngestionClient,
	inputFile: string,
	sampleId: string,
	batchSize: number = 1000,
): Promise<void> {
	console.log(`Parsing GOTTCHA2 data from ${inputFile}`);
	const records = await parseGottcha2TSV(inputFile, sampleId);

	if (records.length === 0) {
		console.error("No valid records found");
		process.exit(1);
	}

	console.log(`Found ${records.length} valid records`);

	// Send in batches
	for (let i = 0; i < records.length; i += batchSize) {
		const batch = records.slice(i, i + batchSize);
		const batchNum = Math.floor(i / batchSize) + 1;
		console.log(`Sending batch ${batchNum} (${batch.length} records)`);

		if (!(await client.sendRecords("ingest-gottcha2", batch))) {
			console.error("Failed to send batch");
			process.exit(1);
		}
	}

	console.log("All records successfully ingested");
}

async function ingestStast(
	client: IngestionClient,
	inputFile: string,
	sampleId: string,
	task: string,
	batchSize: number = 1000,
): Promise<void> {
	console.log(`Parsing STAST data from ${inputFile}`);
	const records = await parseStastTSV(inputFile, sampleId, task);

	if (records.length === 0) {
		console.error("No valid records found");
		process.exit(1);
	}

	console.log(`Found ${records.length} valid records`);

	// Send in batches
	for (let i = 0; i < records.length; i += batchSize) {
		const batch = records.slice(i, i + batchSize);
		const batchNum = Math.floor(i / batchSize) + 1;
		console.log(`Sending batch ${batchNum} (${batch.length} records)`);

		if (!(await client.sendRecords("ingest-stast", batch))) {
			console.error("Failed to send batch");
			process.exit(1);
		}
	}

	console.log("All records successfully ingested");
}

function printUsage() {
	console.log(`
NVD Support Car Ingestion Client

Usage:
  bun run nvd_ingest.ts <command> [options]

Commands:
  gottcha2    Ingest GOTTCHA2 taxonomic abundance data
  stast       Ingest STAST BLAST results
  health      Check service health

Options:
  --url <url>         NVD Support Car service URL (env: NVD_SUPPORT_URL)
  --token <token>     Bearer token for authentication (env: NVD_BEARER_TOKEN)
  --input <file>      Input TSV file
  --sample-id <id>    Sample identifier
  --task <task>       BLAST task type (for stast command, default: megablast)
  --batch-size <n>    Records per batch (default: 1000)
  --help              Show this help message

Examples:
  bun run nvd_ingest.ts gottcha2 --input results.tsv --sample-id SAMPLE001
  bun run nvd_ingest.ts stast --input blast.tsv --sample-id SAMPLE001 --task blastn
  bun run nvd_ingest.ts health
`);
}

// Main entry point
async function main() {
	const args = process.argv.slice(2);

	if (args.length === 0 || args.includes("--help")) {
		printUsage();
		process.exit(0);
	}

	const command = args[0];

	// Parse command line arguments
	const getArg = (name: string, defaultValue?: string): string | undefined => {
		const index = args.indexOf(`--${name}`);
		if (index !== -1 && index + 1 < args.length) {
			return args[index + 1];
		}
		return defaultValue;
	};

	// Get configuration from environment or command line
	const url =
		getArg("url") || process.env.NVD_SUPPORT_URL || "http://localhost:8080";
	const token = getArg("token") || process.env.NVD_BEARER_TOKEN;

	if (!token) {
		console.error(
			"Error: Bearer token is required (--token or NVD_BEARER_TOKEN env var)",
		);
		process.exit(1);
	}

	const client = new IngestionClient(url, token);

	switch (command) {
		case "gottcha2": {
			const inputFile = getArg("input");
			const sampleId = getArg("sample-id");
			const batchSize = parseInt(getArg("batch-size", "1000") || "1000", 10);

			if (!inputFile || !sampleId) {
				console.error(
					"Error: --input and --sample-id are required for gottcha2 command",
				);
				process.exit(1);
			}

			if (!existsSync(inputFile)) {
				console.error(`Error: Input file not found: ${inputFile}`);
				process.exit(1);
			}

			// biome-ignore lint/style/noNonNullAssertion: <Just a script--it's fine if these are null here anyway>
			await ingestGottcha2(client, inputFile!, sampleId!, batchSize);
			break;
		}

		case "stast": {
			const inputFile = getArg("input");
			const sampleId = getArg("sample-id");
			const task = getArg("task", "megablast") || "megablast";
			const batchSize = parseInt(getArg("batch-size", "1000") || "1000", 10);

			if (!inputFile || !sampleId) {
				console.error(
					"Error: --input and --sample-id are required for stast command",
				);
				process.exit(1);
			}

			if (!existsSync(inputFile)) {
				console.error(`Error: Input file not found: ${inputFile}`);
				process.exit(1);
			}

			// biome-ignore lint/style/noNonNullAssertion: <Just a script--it's fine if these are null here anyway>
			await ingestStast(client, inputFile!, sampleId!, task, batchSize);
			break;
		}

		case "health": {
			await client.checkHealth();
			break;
		}

		default:
			console.error(`Error: Unknown command '${command}'`);
			printUsage();
			process.exit(1);
	}
}

// Run the main function
main().catch((error) => {
	console.error("Fatal error:", error);
	process.exit(1);
});

File: examples/scripts/nvd_ingest.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.9"
# dependencies = [
#     "httpx>=0.25.0",
#     "pydantic>=2.0.0",
# ]
# ///

"""
NVD Support Car Ingestion Client

This script provides a command-line interface for sending metagenomic analysis results
to the NVD Support Car service. It supports both GOTTCHA2 and STAST data formats and
handles compression and authentication automatically.

The script uses PEP 723 inline script metadata for dependency management, which means
it can be run directly with uv without requiring a separate requirements file or virtual
environment setup.

Usage with uv:
    uv run nvd_ingest.py --help
    uv run nvd_ingest.py gottcha2 --input results.tsv --sample-id SAMPLE001
    uv run nvd_ingest.py stast --input blast_results.tsv --sample-id SAMPLE001 --task megablast

The script can also be used in Nextflow pipelines by placing it in the bin/ directory
and calling it directly from process scripts.
"""

import argparse
import gzip
import os
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import NoReturn

import httpx
from pydantic import BaseModel, ConfigDict, Field


class Gottcha2Record(BaseModel):
    """GOTTCHA2 taxonomic abundance record"""

    model_config = ConfigDict(populate_by_name=True)

    sample_id: str
    level: str
    name: str
    taxid: str
    read_count: int
    total_bp_mapped: int
    ani_ci95: float = Field(alias="ANI_CI95", ge=0, le=1)
    covered_sig_len: int
    best_sig_cov: float = Field(ge=0, le=1)
    depth: float
    rel_abundance: float = Field(ge=0, le=1)


class StastRecord(BaseModel):
    """STAST BLAST hit record"""

    task: str
    sample_id: str
    qseqid: str
    qlen: int
    sseqid: str
    stitle: str
    length: int
    pident: float = Field(ge=0, le=100)
    evalue: float
    bitscore: float
    sscinames: str
    staxids: str
    rank: str


@dataclass
class IngestionClient:
    """Client for NVD Support Car ingestion endpoints"""

    base_url: str
    bearer_token: str
    timeout: float = 30.0
    verify_ssl: bool = True

    def __post_init__(self) -> None:
        """Initialize HTTP client with configured settings"""
        self.client = httpx.Client(
            timeout=self.timeout,
            verify=self.verify_ssl,
            headers={
                "Authorization": f"Bearer {self.bearer_token}",
            },
        )

    def send_records(self, endpoint: str, records: list[BaseModel]) -> bool:
        """
        Send records to the specified endpoint

        The function converts records to JSONL format, compresses with gzip,
        and sends to the server with proper authentication.

        Returns True on success, False on failure
        """
        # Convert records to JSONL
        jsonl_lines = [record.model_dump_json() for record in records]
        jsonl_data = "\n".join(jsonl_lines).encode("utf-8")

        # Compress with gzip
        compressed_data = gzip.compress(jsonl_data)

        # Send request
        try:
            response = self.client.post(
                f"{self.base_url}/{endpoint}",
                content=compressed_data,
                headers={
                    "Content-Type": "application/gzip",
                    "Content-Encoding": "gzip",
                },
            )
            response.raise_for_status()
            print(f"Successfully ingested {len(records)} records to {endpoint}")
            return True

        except httpx.HTTPStatusError as e:
            print(f"HTTP error {e.response.status_code}: {e.response.text}", file=sys.stderr)
            return False
        except Exception as e:
            print(f"Error sending data: {e}", file=sys.stderr)
            return False

    def close(self) -> None:
        """Close the HTTP client connection"""
        self.client.close()


def parse_gottcha2_tsv(filepath: Path, sample_id: str) -> list[Gottcha2Record]:
    """
    Parse GOTTCHA2 TSV output file

    Expected format has columns:
    LEVEL, NAME, TAXID, READ_COUNT, TOTAL_BP_MAPPED, ANI_CI95,
    COVERED_SIG_LEN, BEST_SIG_COV, DEPTH, REL_ABUNDANCE
    """
    records = []

    with open(filepath) as f:
        # Skip header if present
        header = f.readline().strip()
        if not header.startswith("LEVEL"):
            # No header, reset to beginning
            f.seek(0)

        for line_num, line in enumerate(f, start=2):
            parts = line.strip().split("\t")
            if len(parts) < 10:
                print(f"Warning: Skipping line {line_num}, insufficient columns", file=sys.stderr)
                continue

            try:
                record = Gottcha2Record(
                    sample_id=sample_id,
                    level=parts[0],
                    name=parts[1],
                    taxid=parts[2],
                    read_count=int(parts[3]),
                    total_bp_mapped=int(parts[4]),
                    ani_ci95=float(parts[5]),
                    covered_sig_len=int(parts[6]),
                    best_sig_cov=float(parts[7]),
                    depth=float(parts[8]),
                    rel_abundance=float(parts[9]),
                )
                records.append(record)
            except (ValueError, IndexError) as e:
                print(f"Warning: Error parsing line {line_num}: {e}", file=sys.stderr)

    return records


def parse_stast_tsv(filepath: Path, sample_id: str, task: str) -> list[StastRecord]:
    """
    Parse STAST BLAST output file

    Expected format has columns:
    qseqid, qlen, sseqid, stitle, length, pident, evalue, bitscore, sscinames, staxids, rank
    """
    records = []

    with open(filepath) as f:
        # Skip header if present
        header = f.readline().strip()
        if not header.startswith("qseqid"):
            f.seek(0)

        for line_num, line in enumerate(f, start=2):
            parts = line.strip().split("\t")
            if len(parts) < 11:
                print(f"Warning: Skipping line {line_num}, insufficient columns", file=sys.stderr)
                continue

            try:
                record = StastRecord(
                    task=task,
                    sample_id=sample_id,
                    qseqid=parts[0],
                    qlen=int(parts[1]),
                    sseqid=parts[2],
                    stitle=parts[3],
                    length=int(parts[4]),
                    pident=float(parts[5]),
                    evalue=float(parts[6]),
                    bitscore=float(parts[7]),
                    sscinames=parts[8],
                    staxids=parts[9],
                    rank=parts[10],
                )
                records.append(record)
            except (ValueError, IndexError) as e:
                print(f"Warning: Error parsing line {line_num}: {e}", file=sys.stderr)

    return records


def gottcha2_cmd(args: argparse.Namespace, client: IngestionClient) -> None:
    """Ingest GOTTCHA2 taxonomic abundance data"""
    print(f"Parsing GOTTCHA2 data from {args.input}")
    records = parse_gottcha2_tsv(args.input, args.sample_id)

    if not records:
        print("No valid records found", file=sys.stderr)
        sys.exit(1)

    print(f"Found {len(records)} valid records")

    # Send in batches
    for i in range(0, len(records), args.batch_size):
        batch = records[i : i + args.batch_size]
        print(f"Sending batch {i // args.batch_size + 1} ({len(batch)} records)")

        if not client.send_records("ingest-gottcha2", batch):
            print("Failed to send batch", file=sys.stderr)
            sys.exit(1)

    client.close()
    print("All records successfully ingested")


def stast_cmd(args: argparse.Namespace, client: IngestionClient) -> None:
    """Ingest STAST BLAST results"""
    print(f"Parsing STAST data from {args.input}")
    records = parse_stast_tsv(args.input, args.sample_id, args.task)

    if not records:
        print("No valid records found", file=sys.stderr)
        sys.exit(1)

    print(f"Found {len(records)} valid records")

    # Send in batches
    for i in range(0, len(records), args.batch_size):
        batch = records[i : i + args.batch_size]
        print(f"Sending batch {i // args.batch_size + 1} ({len(batch)} records)")

        if not client.send_records("ingest-stast", batch):
            print("Failed to send batch", file=sys.stderr)
            sys.exit(1)

    client.close()
    print("All records successfully ingested")


def healthz_cmd(args: argparse.Namespace, client: IngestionClient) -> None:
    """Check service health"""
    try:
        response = client.client.get(f"{client.base_url}/healthz")
        response.raise_for_status()
        print(f"Service is healthy: {response.text}")
    except Exception as e:
        print(f"Service health check failed: {e}", file=sys.stderr)
        sys.exit(1)
    finally:
        client.close()


def main() -> None:
    """Main CLI entry point"""
    parser = argparse.ArgumentParser(
        description="NVD Support Car ingestion client",
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )

    # Global options
    parser.add_argument(
        "--url",
        default=os.getenv("NVD_SUPPORT_URL", "http://localhost:8080"),
        help="NVD Support Car service URL (env: NVD_SUPPORT_URL)",
    )
    parser.add_argument(
        "--token",
        default=os.getenv("NVD_BEARER_TOKEN"),
        required=not os.getenv("NVD_BEARER_TOKEN"),
        help="Bearer token for authentication (env: NVD_BEARER_TOKEN)",
    )
    parser.add_argument(
        "--no-verify-ssl",
        action="store_true",
        help="Disable SSL certificate verification",
    )

    # Subcommands
    subparsers = parser.add_subparsers(dest="command", help="Available commands")

    # GOTTCHA2 command
    gottcha2_parser = subparsers.add_parser(
        "gottcha2",
        help="Ingest GOTTCHA2 taxonomic abundance data",
    )
    gottcha2_parser.add_argument(
        "--input",
        "-i",
        type=Path,
        required=True,
        help="GOTTCHA2 TSV file",
    )
    gottcha2_parser.add_argument(
        "--sample-id",
        "-s",
        required=True,
        help="Sample identifier",
    )
    gottcha2_parser.add_argument(
        "--batch-size",
        type=int,
        default=1000,
        help="Records per batch (default: 1000)",
    )

    # STAST command
    stast_parser = subparsers.add_parser(
        "stast",
        help="Ingest STAST BLAST results",
    )
    stast_parser.add_argument(
        "--input",
        "-i",
        type=Path,
        required=True,
        help="STAST BLAST output file",
    )
    stast_parser.add_argument(
        "--sample-id",
        "-s",
        required=True,
        help="Sample identifier",
    )
    stast_parser.add_argument(
        "--task",
        "-t",
        default="megablast",
        help="BLAST task type (e.g., megablast, blastn) (default: megablast)",
    )
    stast_parser.add_argument(
        "--batch-size",
        type=int,
        default=1000,
        help="Records per batch (default: 1000)",
    )

    # Health check command
    healthz_parser = subparsers.add_parser(
        "healthz",
        help="Check service health",
    )

    # Parse arguments
    args = parser.parse_args()

    # Check if command was specified
    if not args.command:
        parser.print_help()
        sys.exit(1)

    # Check required token
    if not args.token:
        print(
            "Error: Bearer token is required (use --token or set NVD_BEARER_TOKEN env var)",
            file=sys.stderr,
        )
        sys.exit(1)

    # Create client
    client = IngestionClient(
        base_url=args.url.rstrip("/"),
        bearer_token=args.token,
        verify_ssl=not args.no_verify_ssl,
    )

    # Execute command
    if args.command == "gottcha2":
        gottcha2_cmd(args, client)
    elif args.command == "stast":
        stast_cmd(args, client)
    elif args.command == "healthz":
        healthz_cmd(args, client)


if __name__ == "__main__":
    main()
